File: projects/Python/examples/serialization_final.py
import fbe
import proto
from proto import proto


def main():
    # Create a new account with some orders
    account = proto.Account(1, "Test", proto.State.good, proto.Balance("USD", 1000.0), proto.Balance("EUR", 100.0))
    account.orders.append(proto.Order(1, "EURUSD", proto.OrderSide.buy, proto.OrderType.market, 1.23456, 1000.0))
    account.orders.append(proto.Order(2, "EURUSD", proto.OrderSide.sell, proto.OrderType.limit, 1.0, 100.0))
    account.orders.append(proto.Order(3, "EURUSD", proto.OrderSide.buy, proto.OrderType.stop, 1.5, 10.0))

    # Serialize the account to the FBE stream
    writer = proto.AccountFinalModel(fbe.WriteBuffer())
    writer.serialize(account)
    assert writer.verify()

    # Show the serialized FBE size
    print("FBE size: {}".format(writer.buffer.size))

    # Deserialize the account from the FBE stream
    reader = proto.AccountFinalModel(fbe.ReadBuffer())
    reader.attach_buffer(writer.buffer)
    assert reader.verify()
    reader.deserialize(account)

    # Show account content
    print()
    print(account)


if __name__ == "__main__":
    main()

File: projects/Swift/Tests/Examples/SerializationFinal.swift
//
//  SerializationFinal.swift
//

import XCTest
import ChronoxorProto

class ExampleSerializationFinal: XCTestCase {
    func testSerializationFinal() {
        // Create a new account with some orders
        var account = Account(id: 1, name: "Test", state: State.good, wallet: Balance(currency: "USD", amount: 1000.0), asset: Balance(currency: "EUR", amount: 100.0), orders: [])
        account.orders.append(Order(id: 1, symbol: "EURUSD", side: OrderSide.buy, type: OrderType.market, price: 1.23456, volume: 1000.0))
        account.orders.append(Order(id: 2, symbol: "EURUSD", side: OrderSide.sell, type: OrderType.limit, price: 1.0, volume: 100.0))
        account.orders.append(Order(id: 3, symbol: "EURUSD", side: OrderSide.buy, type: OrderType.stop, price: 1.5, volume: 10.0))

        // Serialize the account to the FBE stream
        let writer = AccountFinalModel()
        _ = try! writer.serialize(value: account)
        assert(writer.verify())

        // Show the serialized FBE size
        print("FBE size: \(writer.buffer.size)")

        // Deserialize the account from the FBE stream
        let reader = AccountFinalModel()
        reader.attach(buffer: writer.buffer)
        assert(reader.verify())
        _ = reader.deserialize(value: &account)

        // Show account content
        print("")
        print(account.description)
    }
}
